client/state/themes/actions.js

```diff
@@ -404,42 +404,34 @@ export function clearActivated( siteId ) {
 }
 
 /**
- * Triggers a network request to activate a specific wpcom theme on a given Jetpack site.
- * First step of the process is the installation of the theme on Jetpack site.
- * Second step is the actuall activation.
+ * Triggers a network request to install and activate a specific theme on a given
+ * Jetpack site. If the themeId parameter is suffixed with '-wpcom', install the
+ * theme from WordPress.com. Otherwise, install from WordPress.org.
  *
- * Warning: this is working but not final version. It has a built in hack in form of
- * dispatching THEME_ACTIVATE_REQUEST twice to mark process start and has no handlin
- * of information about installation process except in failure scenario.
- *
- * @param  {String}   themeId   Theme ID, this should be standard id without -wpcom suffix.
+ * @param  {String}   themeId   Theme ID. If suffixed with '-wpcom', install theme from WordPress.com
  * @param  {Number}   siteId    Site ID
  * @param  {String}   source    The source that is reuquesting theme activation, e.g. 'showcase'
  * @param  {Boolean}  purchased Whether the theme has been purchased prior to activation
  * @return {Function}           Action thunk
  */
-export function activateWpcomThemeOnJetpack( themeId, siteId, source = 'unknown', purchased = false ) {
-	//Add -wpcom suffix. This suffix tells the endpoint that we want to
-	//install WordPress.com theme. Without the suffix endpoint would look
-	//for theme in .org
-	const suffixedThemeId = themeId + '-wpcom';
-	return dispatch => {
+export function installAndActivate( themeId, siteId, source = 'unknown', purchased = false ) {
+	return ( dispatch ) => {
+		// To the user, this is presented as just activation -- the install part is hidden.
+		// Thus, we want to trigger any UI changes that apply to activation here, too.
 		dispatch( {
 			type: THEME_ACTIVATE_REQUEST,
-			themeId: suffixedThemeId,
+			themeId,
 			siteId,
 		} );
-
-		return wpcom.undocumented().installThemeOnJetpack( siteId, suffixedThemeId )
+		return dispatch( installTheme( themeId, siteId ) )
 			.then( () => {
-				return activateTheme( suffixedThemeId, siteId, source, purchased )( dispatch );
+				dispatch( activateTheme( themeId, siteId, source, purchased ) );
 			} )
-			.catch( ( error ) => {
+			.catch( () => {
```

Comment:
Hmm, good point. I think in that case I'm also going to remove the `THEME_ACTIVATE_REQUEST` [dispatch](https://github.com/Automattic/wp-calypso/pull/10161/files/0f7c45bd92c6f737f5f23c157ccea51cb2095818#diff-dd8c5284a85f9baf309a0ec3a482d228L428) because it'd be otherwise way too confusing (lack of expected symmetry etc). Instead, we should probably have the UI listen to a separate `isInstalling` state.